From a60576b1cf347830d515db8cc5d46c004f8f1ede Mon Sep 17 00:00:00 2001
From: Jacob McSwain <jacob.a.mcswain@gmail.com>
Date: Thu, 27 Oct 2016 23:18:52 -0400
Subject: [PATCH 4/6] OMS-N: Allow on-demand window refreshes using "om
 refresh"

This commit allows Substratum to call "om refresh" to trigger window
refreshes on demand.

Original implementation for M by @USA-RedDragon
Current and further development by @nicholaschum

Signed-off-by: sub77 <sub77@ymail.com>
---
 .../server/om/OverlayManagerShellCommand.java      | 28 ++++++++++++++++++++++
 1 file changed, 28 insertions(+)

diff --git a/services/core/java/com/android/server/om/OverlayManagerShellCommand.java b/services/core/java/com/android/server/om/OverlayManagerShellCommand.java
index 44004c1..b9abece 100644
--- a/services/core/java/com/android/server/om/OverlayManagerShellCommand.java
+++ b/services/core/java/com/android/server/om/OverlayManagerShellCommand.java
@@ -64,6 +64,8 @@ final class OverlayManagerShellCommand extends ShellCommand {
                 	return runDisableAll();
                 case "set-priority":
                     return runSetPriority();
+                case "refresh":
+                    return runRefresh();
                 default:
                     return handleDefaultCommands(cmd);
             }
@@ -100,6 +102,8 @@ final class OverlayManagerShellCommand extends ShellCommand {
         out.println("    'lowest', change priority of PACKAGE to the lowest priority.");
         out.println("    If PARENT is the special keyword 'highest', change priority of");
         out.println("    PACKAGE to the highest priority.");
+        out.println("  refresh [--user USER_ID]");
+        out.println("    Refresh all overlay packages.");
     }
 
     private int runList() throws RemoteException {
@@ -266,4 +270,28 @@ final class OverlayManagerShellCommand extends ShellCommand {
             return mInterface.setPriority(packageName, newParentPackageName, userId) ? 0 : 1;
         }
     }
+
+    private int runRefresh() {
+        int userId = UserHandle.USER_SYSTEM;
+        int ret = 1;
+        try {
+            String opt;
+            while ((opt = getNextOption()) != null) {
+                switch (opt) {
+                    case "--user":
+                        userId = UserHandle.parseUserArg(getNextArgRequired());
+                        break;
+                    default:
+                        System.err.println("Error: Unknown option: " + opt);
+                        return 1;
+                }
+            }
+            mInterface.refresh(userId);
+            ret = 0;
+        } catch(Exception e) {
+            e.printStackTrace();
+            ret = 1;
+        }
+        return ret;
+    }
 }
-- 
2.1.4

